### PR TITLE
Add a tile manager to reduce number of requests to firestore

### PR DESCRIPTION
**Pull request by sean-prz**
_closes #60_
### What
This PR aims to reduce drastically the number of reads to the Firestore database, by focusing on the request for parking spots on the map screen.
### Why
The number of READS to the database are getting close to the free quota limit, because of the following reasons:   
- The Data is reloaded each time the map is reloaded (when navigating to the map)
- The Data is also reloaded when we un-zoom the map. 
- The Data is reloaded if we scroll to another place and back.

All of this useless re-fetch should be avoided with this PR

### How
This PR introduces a Tile manager, that remember which portion of the map has already been loaded.

The MapScreen doesn't change.
The Map screen still request about three times the portion of the map to make the experience smooth to the viewmodel.
However the zone requested is segmented in Tile of 0.1 by 0.1 degrees and the viewmodel check if these tiles have already been loaded, before requesting it to the db.
In this sense the List of parkings stored in the viewmodel is not overwritten at each call anymore but rather appended with the new result of the new tiles loaded.

**Edit** : This PR now also make sure to only display relevants spots, while still caching the data even if not rendered by the UI

**Edit 2** This PR also prevents the user from flooding the database with request when the map is unzoomed.
If the map is unzoomed too much no request is passed and no spots are displayed.

### Test
The current tests for the viewmodel and End2End are used to check that the Tile manager  implementation did not break anything. 

-- -
This Big fix introduces some new issuses that need to be addressed #137 and #136

image

### CODE REVIEW

**Comment by matttsss** [comment]
I'm not sure this implementation is very friendly with the UI thread. In this configuration, the State-Flow passed to the UI thread would hold way to many parkings... It might be a better idea to have the Tiles hold their parkings and on `getParkingsInRect` to pass only the necessary parkings from the Tile set.
In addition, I am not sure that a `MutableList` is the best choice of data-structure for this since we don't need ordering, if you could make use of a `LinkedHashMap` like we did in Javions. This adds the benefit of putting a limit to the cached memory.

**Comment by sean-prz** [comment]
> I'm not sure this implementation is very friendly with the UI thread. In this configuration, the State-Flow passed to the UI thread would hold way to many parkings... It might be a better idea to have the Tiles hold their parkings and on `getParkingsInRect` to pass only the necessary parkings from the Tile set. In addition, I am not sure that a `MutableList` is the best choice of data-structure for this since we don't need ordering, if you could make use of a `LinkedHashMap` like we did in Javions. This adds the benefit of putting a limit to the cached memory.

Agree, I made the changes you requested. The UI now only gets what it needs, while the viewmodel still keep the cached data

**Review by matttsss — COMMENTED** [review]
Pretty nice changes, one extra thing we could do later is not reload to check if a tile was already in view but this is a nice change and should help with the Firebase queries.
One last thing, I was thinking we could maybe try and initialize our `LinkedHashMap` like we did in `Javions`: 
image
We could try and overload that function to avoid having to big of a cache

**Comment by sean-prz** [comment]
> Pretty nice changes, one extra thing we could do later is not reload to check if a tile was already in view but this is a nice change and should help with the Firebase queries. One last thing, I was thinking we could maybe try and initialize our `LinkedHashMap` like we did in `Javions`: image We could try and overload that function to avoid having to big of a cache

Okay, I just pushed a fix to init our hashMap with some custom parameters.
 Let's hope our app becomes big enough that these optimizations problems become relevant !

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt [hunk 1]
```diff
@@ -97,13 +105,24 @@ class ParkingViewModel(
...
+        return@forEach // Skip to the next tile if already fetched
+      }
+      tilesToParking.value[tile] = emptyList() // Avoid querying the same tile multiple times
+      parkingRepository.getParkingsBetween(
+          tile.bottomLeft,
+          tile.topRight,
+          { parkings ->
+            tilesToParking.value[tile] = parkings
+            _rectParkings.value +=
+                tilesToParking.value.filter { it.key in tilesToDisplay }.values.flatten()
```
Comment (matttsss) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:122]: This increment value could just be `parkings` right?

Comment (sean-prz) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:122]: ooooooh that is so much faster and niceer, I feel dumb, I am gonna try

**Review by matttsss — COMMENTED** [review]
The parkings are well displayed and the UI is fluid, however I have a little issue while adding parkings... They don't appear once I add them :sweat_smile:

**Comment by sean-prz** [comment]
> The parkings are well displayed and the UI is fluid, however I have a little issue while adding parkings... They don't appear once I add them 😅

Nice catch I forgot to test that. It makes sense as the map is cached. We could  force the cache for this tile to be reloaded but this seems overkill. I think this issue need more time to be think about, I will create an issue and make a PR to fix this later on

**Review by matttsss — APPROVED** [review]

**Review by Armetiis — COMMENTED** [review]
I don't have a lot to add thanks to @matttsss 😂. This PR clearly progressed a lot. Thanks for the much needed changes!

**Review by Armetiis — APPROVED** [review]

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt [hunk 2]
```diff
@@ -197,7 +198,10 @@ fun MapScreen(
                     loadedBottomLeft = loadedCorners.first
                     loadedTopRight = loadedCorners.second
 
-                    parkingViewModel.getParkingsInRect(loadedBottomLeft, loadedTopRight)
+                    // Temporary fix to avoid loading too much parkings when zoomed out
+                    if (mapView.mapboxMap.cameraState.zoom > thresholdDisplayZoom) {
```
[the remark this answers is not in the document]
Comment (sean-prz) [reply app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt]: Done !

File: app/src/main/res/drawable/dot.png
[the remark this answers is not in the document]
Comment (sean-prz) [reply app/src/main/res/drawable/dot.png:1]: Accident happens

**Pull request merged**